### ISSUE TITLE
LokiDB op organization and binary index strategies

### ISSUE THREAD

**Issue by obeliskos** [LokiJS-Forge/LokiDB#95]
## I'm submitting a...

<pre><code>
[ ] Regression (a behavior that used to work and stopped working in a new release)
[ ] Bug report  
[ ] Feature request
[ ] Documentation issue or request
[ ] Support request => Please do not submit support request here, instead see https://github.com/LokiJS-Forge/LokiDB/blob/master/CONTRIBUTING.md#question
[x] Refactoring Discussion?
</code></pre>

In LokJS, our ops 'evolved' when adding binary indices and to adapt to problems with javascript equality and relational ops when dealing with dirty data and mixed types.  Recently we have even added ops for explicitly opting for js relational ops ($jgt, $jgte, $jlt, $jlte, $jbetween)... not for functional output differences but just because they are faster in some situations.   

I also suggest me might add different 'modes' for dealing with relational operations... by default we should probably continue to expect that users will throw dirty data at it and we will need to rely on 'loki sorting' to establish a reliable 'range' over which our binary indices can work.   But many people may be producing their own data internally and have confidence over the cleanliness of their data... especially if they also adopt typescript for code that creates documents they insert.... in those cases we might want to 'reroute' our $ltHelper and $gtHelper to utilize native javascript relational and equality operators... doing so should increase performance even more.

I will propose suggestions later but for now I will just attempt to quantify/itemize factors involved in this for discussion.

**Comment by Viatorus** [comment]
Thank you obeliskos, I thought about similar things. I am also open to any breaking changes to LokiJS (which already exist). That's way LokiDB is still in beta.

>  by default we should probably continue to expect that users will throw dirty data at it 

I would argue against this. LokiDB is for programmer/software developers and not for "script kiddies". I know most of the time which data I expect when dealing with databases. As you already said, TypeScript helps there a lot. ;)

I would prefer strict comparison should be the default behavior.

Talking about different modes for binary indices one could add it while creating a collection, like so:

```js
const coll = db.addCollection("CoolStuff", {indices: ["id", {field: "name", strict: false}]}
```
Where the field `id` would use strict comparision on binary indices and the `name` field abstract/weak comparision.

Additionally, I would rename some operations so that _$eq, $ne, $lt, $lte, $gt, $gte, $between_ would use the javascript strict comparision and _$aeq, $ane, $alt, $alte, $agt, $agte, $abetween_ use weak comparision.

**Comment by obeliskos** [comment]
It may take a lot of effort but for that level of granularity, I think we should look into extracting as much binary index logic as possible and adding modes to that.  I was originally thinking of 'db'-level option but collection or property (index-level) modes would be flexible.

Dates are another example which (unless we encode them ourselves) would need non-strict indices to account for the fact they are objects (never strict equal) but <, > will convert to string before doing string relational tests.  Not to mention they cannot be serialized without being converted to string.

Also to reword my comments to be more accurate, javascript doesn't have a strict relational comparisons unless you test typeof() yourself. But for most types (other than Dates), sorting can work because ===, <, and > will all 'agree'.

**Comment by obeliskos** [comment]
I am thinking it will be its own ts class (binary_index.ts?).

I will need to add an internal array of values so that it doesn't need to 'ask' collection for those.  Might also optimize translation between data[] positions and sorted index[] positions and/or $loki id values.

The should really fix javascript dates in future js standard, i only deal with unix ms format within my data sets because of that.

**Comment by Viatorus** [comment]
Hey obeliskos,
any progress on this? 😃

**Comment by obeliskos** [comment]
Nope but your timing is good and I'll start looking at it this weekend. 

- I may end up pursuing a full binary tree class implementation rather than just binary index... at least at first.  Btree should use more memory but speed insertions and removals which happen even during updates.
- Our $loki index can probably be converted to a Unique index or Map.
- Our existing Unique Index might be rewritten to avoid (if possible) nested object references (maybe value references instead?)
- I will probably need to at least borrow lokijs benchmarks temporarily to ensure parity of performance baseline to lokijs and then with my changes.
- I will begin work on a new binary_index branch

**Comment by obeliskos** [comment]
I have done most of the 'last mile' integration of binary tree index into loki.  Since I am new to typescript and alot of the loki ts refactoring, feel free to criticize or suggest changes to better support long term maintenance.  Once integrated you can still change at will if you want or are interested.

### Overview of changes
To summarize changes, BinaryTreeIndex is self balancing (AVL) tree implementation which allows duplicate values (stored internally as siblings).

In order to integrate BinaryTreeIndex into loki, I had to refactor comparators and add several interfaces to helper.js... we may eventually want to rename helper.ts into something like ranged_compare.js.

I also wanted to allow user to inject their own comparators and rangedIndex types (should they want to).  So loki will maintain a map of named comparators and rangeIndex factory methods.  Currently they defined globally within helper.js but a possibility might be that we store them in loki instance and dependency inject them into collections... that is if we want to go with a DI approach.

### Example integration usage : 
```javascript
let db = new Loki("test.db", {
  comparatorMap : {
    numberCompare: { compare(a,b)=> return a-b; } // js implementation of ILokiRangedComparer
  },
  rangedIndexFactoryMap: {
    redblack: (name, comparator) => {
      return new MyRedBlackIndexImplementation(name, comparator); // defined elsewhere in user code
    }
  }
});

// loki will register "js" and "loki" comparators into (global) ComparatorMap
// loki will register "btree" indexTypeName and factory function into (global) RangedIndexFactoryMap

let coll = db.addCollection("coll", {
  rangedIndexes: [
    {
      age: { indexTypeName: "btree", comparatorName: "numberCompare" }
      address: { indexTypeName, "bidx", comparatorName: "js" } // eventually we may convert to rangedindex
      anotherDirtyProp: { indexTypeName: "redblack", comparatorName: "loki" }
      dirtyData: {}, // implicit indexTypeName: bidx? implicit comparatorName : "loki"?
    }
  ]
});
```
### $loki -vs- data array positions
So currently the RangedIndex will be implemented as key/value with key=$loki (number) and val as generic.  As such, when performing find() operations i have to look up each document with a get() to get each document reference or data aray position to add to result.  For now the $loki index lookups seem to be adding low overhead.  I have thought about replacing resultset filtered rows to contain $loki id's instead of data array positions, but this would probably be done along with either a conversion of old binary index to IRangedIndex implementation, or removal of old binaryIndices now that there is a replacement.  BinaryTreeIndex has slightly more memory overhead than old binary indices but i think they may not be necessary.

### Benchmarks
Initial benchmark results show that BinaryTreeIndex is about as fast (maybe a little faster) than the old binary indices when find() ops are taken on their own.  When index maintenance operations (insert, update, remove) are accounted for, BinaryTreeIndex is much faster.  In order to more accurately measure performance, the benchmarks need to be invoked with --expose-gc command line parameter so that i can garbage collect memory in between benchmarks.  There are still a few anomalies i need to look into, but they seem to show BinaryTreeIndex scaling much better to higher document counts.  Currently, i am not checking in the built packages which they import, so you might need to do a build before running them.

### loki idIndex and unique index
Since the overhead of performing a get() seems low, I am thinking i will not replace the loki idIndex with hashmap similar to unique index for now.  If you have different opinion or need for this, let me know.  One anomaly in the above benchmarks is that loki get() seems to be faster than unique by() when garbage collection is performed in between benchmarks.  If that proves the case, I may even suggest we convert unique index to be simple key/value of (propertyValue/$loki) and do a get() to resolve.  Since our current Unique Index are not serialized and requires recreation load time (based on persisted column names), we could eliminate that load time delay by serializing the index itself and just getting rid of the 'uniqueNames' from Serialized interface.

### Versioning / upgrade support
I have not done anything to version 'upgrade' the database.  It should be backwards compatible... if no ranged indexes were saved they would just be empty until populated.  Should we decide to get rid of older binary indices and replace that with conversion to BinaryTreeIndex, we might add need to add that conversion logic to upgrade codepath.

### Unit testing
I have been implementing code coverage for most of my additions but i have had somewhat limited visibility as to its accuracy or what travis rules might require before complaining.  Perhaps during a merge PR i will be able to see before/after stats a little more clearly.  Let me know if you foresee any issues which might be encountered.

### Feedback
So depending on your interest and available time to look into this, I am interested in whether you think the global ComparatorMap and RangedIndexFactoryMap should be organized or refactored differently.  
If you see any areas where I have not implemented Typescript in the best manner, feel free to point that out... or any over questions or concerns before i begin looking into merging via pr.

After doing some preliminary tests on separate repo, I am somewhat dreading a rebase from master before merge PR, so I might tread lightly with experimentation of vs code conflict resolution.  In my test I had to resolve conflicts leading to headless parent i had to publish to new branch before merge pr.

**Comment by obeliskos** [comment]
I did a rebase and it went smoothly so I went ahead and created a pull request to see what travis would say.  Appears to have less code coverage than optimal... I will have to analyse to see if it points out where I might add more tests.  

I won't rush the pull request in case you have interest in providing feedback... otherwise I may look into adding a few unit tests and merge later next week.

**Comment by Viatorus** [comment]
Thank you for your good summary! Now I have to understand what you are doing. ;)

I don't have much time this week so please, be patient with me.

**Comment by obeliskos** [comment]
No problem, no rush, and if you don't want to know all details, you might just make sure the interface is clean enough (helper.ts has most interfaces).

**Comment by Viatorus** [comment]
Hello obeliskos,

I have some first suggestions (just before I go on holidays for 4 days ;)).

The stuffed shirt in me: I improved our TSLint configuration to better standardize the code format.
Please rebase your branch and with `npm run lint:fix` most, if not all the linter errors should be fixed for you. ;)

Second, I would suggest to drop the old binary index support. Your BinaryTree is much cleaner and not directly integrated into the Collection.

As upgrading support from LokiJS to LokiDB I only see to reindex with BinaryTrees - this would be done only once.

**Comment by obeliskos** [comment]
Thanks, good time for holidays... hope you enjoy yours :)

Thanks, I'll start that later today along with examining test coverage.

Early on, I was having issues with getting indexers to work with (string|number) so i hardcoded id (key) to be number ($loki) and just had a generic for value type.   It seems you are able to do this with Dict<InvertedIndex> so i may look into experimenting with various syntax to see if I can allow either strings or numbers for the hash object key.

**Comment by Viatorus** [comment]
Hey @obeliskos,
thank you for your ongoing work. But we have to revert (reset) the last two commits...
Our changelog gets automatically generated from the commit messages if they follow the [Commit Message Guidelines](https://github.com/LokiJS-Forge/LokiDB/blob/master/CONTRIBUTING.md#-commit-message-guidelines).

I backed up both commits at _backup/btree_index_feature_ and _backup/remove_legacy_binary_index_.

If you are okay with it, I will git force back two commits. After that, you can create both PRs again and I would merge them. I am sorry.

**Comment by Viatorus** [comment]
Furthermore, I would like to adjust the API a little bit and naming. But not before the commits are correct.

**Comment by Viatorus** [comment]
Hey @obeliskos,
So I finished the commit messages issue after messing it by myself. :fallen_leaf: 

I saw this commit message:
> still need to determine best way to (re-implement) having indexed property on property with dot notation (nested prop)

I think this is obsolete. With the `nestedProperties` options you can use dot notations.

**Comment by obeliskos** [comment]
great, thanks :)  

The only other -index- infrastructure change i was thinking we should make is to UniqueIndex...  We currently do not save unique indexes within serialized database but just an array of property names that we will regenerate unique indexes upon deserialization/load.   

I was considering converting it into a very basic key value lookup with 'val' being the key and $loki being the value.  An similar (second) key value lookup might be made from 'val' to the actual object instance.  Then we might strip out the second kvp within toJSON (saving the actual index within db serialization), and on subsequent loads, the -first- by() access would lazy lookup the instance by $loki (which is nearly as fast).

I first intend to experiment to try to get an understanding as to why .get() ops seem faster than .by() ops in my benchmarks as that should not really be the case.  

I am not sure we are at a point to consider converting loki.data from an array to a hashmap or whether that would be worth it.  

And as for converting resultset.filteredrows into array of $loki ids rather than data[] positions, I am not sure I am ready for that either at the moment.  We can take time to consider those changes.

Thanks for info about nestedProperties, not sure if and how to ensure test coverage of that.

Feel free to suggest API and naming adjustments.

**Comment by Viatorus** [comment]
Feel free to change the internal data structure if you feel this would be better.
We don't have a release date yet so don't feel rushed.

I like the option for the user, that he can decide if the index should be stored or not (like in the full-text search option). Not all indexed values can be serialized.

I try to get some stuff done for the new binary tree tonight (UTC+2h).

**Comment by obeliskos** [comment]
I have created a new branch for unique index and lokimap optimizations.  

I see in unique index (you?) removed the object references which is probably what i would have done.  I added second map so we have two maps (loki->val) and (val->loki).

I also added a collection level 'lokimap' which maps (loki->obj) with obj being a data array object reference.

I modified collection.by() to directly use the index rather than use slower find() codepath.

I modified collection.get() to use lokimap whenever we do not pass true for second param to return data array position.  Hopefully soon we can rewrite resultset.filteredrows to use $loki and lokimap rather than data array positions.

collection.by is now as fast as it was in lokijs and collection.get is faster (same perf as by()) when not requesting data array position (same perf as before if requesting dap as that path uses binary search)

The only thing i have left is to look into whether we should persist the unique index in the database or regenerate it every time... I am typically of the mind to persist them in database but not sure if you agree.  Or should we allow user to specify whether to serialize indexes.

**Comment by Viatorus** [comment]
> I see in unique index (you?) removed the object references 
Yes, I did. ;)

>  so we have two maps (loki->val) and (val->loki)
I don't see any gain from this. Why do we need loki -> val?

Do you plan to refactor `filteredRows`, so that we only us `$loki` as only data position identifier?

> The only thing i have left is to look into whether we should persist the unique index in the database or regenerate it every time... I am typically of the mind to persist them in database but not sure if you agree. Or should we allow user to specify whether to serialize indexes.

Let this to me. I will add this functionality (also for rangedIndex) later.

**Comment by obeliskos** [comment]
Why have two maps?  Well, for one example to get rid of the iteration we needed to do in the update.  From an abstract perpective it enforces 1-1 mapping, whereas a single hashmap would seem to enforce only many-1 mapping.  Remove accepts $loki and can easily computed the 'cached value' (whether cloning was enabled or not) which acts as a directly hash index for the other map, and vice versa for updates or inserts.

I was intending to use lokimap (not only for performance boost) but ideally, yes to eventually replace usage (such as in filteredrows), to using $loki instead.  Once all codepaths no longer rely on data array position we could convert collection.data to map itself and get rid of lokimap.  Certain areas such as iteration performance (object.keys) for unindexed ops remains to be seen, but converting to $loki should simplify things.

Having said that, do you have any issues with my merging this into master?

**Comment by obeliskos** [comment]
Actually, it would probably be better to not pass in doc and have all method params be either loki (number) or val (any).  The doc probably will never matter if we do not cache object reference.

**Comment by Viatorus** [comment]
> Once the above is resolved, I would probably next do (simple) pr to rename "btree" and "btree-index.ts" to "avl" and "avl-index.ts" to avoid confusion in the future... I might rename BinaryTreeIndex class itself to AvlTreeIndex. Those changes should be simple but intersect a few files.

I like it to be specific, since it is nott a B-tree. ;)

> If you depend on (loki) data array positions within your inverted index packages/classes, let me know. Whichever way we decide to go for collection.data for a non-beta release, all other above changes still have merits of their own.

Not at all. Currently I use the array index but $loki would also work.

---

This is a huge task you have planned! Good luck! I am happy over every PR.

**Comment by obeliskos** [comment]
"What, if anything, the B stands for has never been established." :)

Thanks

**Comment by obeliskos** [comment]
For my own personal knowlege, and since I have made considerable changes to the benchmarks from LokiJS to LokiDB, I wanted to test them head-to-head with the same benchmark, so I did a test of that... requiring both libraries and switching between them to compare.

For the most part unique/get and unindexed finds were very similar, so much so I wont bother posting.

Where i was really interested was difference between LokiDB/AVL and LokiJS/BinaryIndices (adaptive)... here are the results so we can mentally file them before moving on : 

**Finds (Low Document Count - 20K) :** 
(BI) 716,537 ops/s => (AVL) 1,096,431 ops/s

**Finds (High Document Count - 200K) ** 
(BI) 266,152 ops/s => (AVL) 1,086,366 ops/s

**Nightmare/Maintenance (Low Document Count - 40K)**
insert+find : (BI) 55,286 ops/s => (AVL) 216,837 ops/s
remove+find : (BI) 23,702 ops/s => (AVL) 270,268 ops/s
update+find : (BI) 127,481 ops/s => (AVL) 201,116 ops/s

**Nightmare/Maintenance (High Document Count - 100K)**
insert+find : (BI) 15,780 ops/s => (AVL) 235,737 ops/s
remove+find : (BI) 6,218 ops/s => (AVL) 310,231 ops/s
update+find : (BI) 7,099 ops/s => (AVL) 205,482 ops/s

**Comment by Viatorus** [comment]
So LokiDBs AVL outperformance LokiJS' binary index. :)

That's great news and also it is perfect for advertising LokiDB once it is released. ;)

**Comment by Viatorus** [comment]
Thank you for thinking this through.

I hope the user don't get to confused. Maybe we can adopt some naming convention from other databases.

Overall great idea!

**Comment by obeliskos** [comment]
With the latest commit 05fa442 I think I am mostly complete with the feature and refactorings.

To summarize, in loki constructor the user can inject their own : 
- comparators
- ranged index implementations
- loki operator packages (collection of ops, each of which is like LokiOps)

Within collection, user can define : 
- unindexedSortComparator
- defaultLokiOperatorPackage (string/name to above hashmap) to use that package within its find() ops

In the future i can see our resultset.find method accepting an options object rather than just bool firstOnly.  In that find object, we might allow user to specify override for ops package.

We will probably need 'wiki-like' help for some of the higher level documentation of items such as this.  I am thinking maybe we should do this within markup within project (and linked to from lokidb wiki page). Any objection to doing that, and if not, where in our folder structure might we do that so as to not conflict with existing /docs folder which I believe will be used for autogenerated documentation reference?

I have planned to add and clean up unit test coverage around this, but if you think you will be wanting to make your own changes/reorganizations soon let me know and i will pr this and change later.  Having two people make massive changes at this stage can be cumbersome :S If not, I hope to clean up and p.r. this sometime next week.

**Comment by Viatorus** [comment]
> We will probably need 'wiki-like' help

Please use the docs folder for this. We are using [mkdocs](https://www.mkdocs.org/) for documentation. Create a new markdown file and write your documentation. Each release updates the documentation. :)

For example, I have this documenation in the pipeline:
https://github.com/LokiJS-Forge/LokiDB/blob/docs/docs/full-text-search.md

**Comment by Viatorus** [comment]
Hey obeliskos,
Thank you. I will review it tonight (UTC+2).

**Comment by obeliskos** [comment]
Thanks for the review @Viatorus...

A few suggestions about how I think we should handle of pr's in the future :

- I don't think we should necessarily -require- code reviews for collaborators. That feature was recently turned on, but I don't think it should be.  My last few pr's have been fairly significant, so it makes sense for you to ask to, or me to ask if you would want to, review pr's like these (if your time permits) to ensure we are agreement of major changes and additionally to find and question potential bugs, best practices, typos, etc.   Perhaps if we get 5 or 10 active contributors, I can see requiring code reviews being a good policy requirement, but for just us I hope we know what the other has concerns about and might want to review.  

- If either of us would like to review pull requests, we can clarify which or all of these we would like to review and the other can assign and await the review for those pr's.  If either of us are to provide a code review, we should just ask questions, request changes, or approve. As collaborators, we should probably all be committing our own pull requests. 

I think we've both made good efforts to collaborate better with each other's workflow.  If there are issues with the above or any remaining areas you would like to see improved, feel free to discuss here or in another issue.

I'll close out this issue soon, as the scope I had for it has been achieved.  As such, the -major- changes which I had considered 'required for v1.0' have been implemented. 

Let me know if you think it would be a good idea for us to create a new (or use an existing) issue to itemize any remaining changes we would like see make it into a version 1 release.  That might allow ourselves and others can better gauge our progress towards that.

**Comment by Viatorus** [comment]
I hope you didn't take this personal after I enabled the review requirements. I just thought this is a good feature to keep track on code changes for both of us. Sorry that I merged your PRs after review. This should be as you said the part of the author and not reviewer.

I will undo it.

**Comment by obeliskos** [comment]
No worries... thanks 😃
